File: java_projects/glue_code/commons-graph/src/main/java/org/apache/commons/graph/shortestpath/ShortestDistances.java
package org.apache.commons.graph.shortestpath;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import org.apache.commons.graph.ContextInitializer;
import org.apache.commons.graph.weight.OrderedMonoid;
import org.graalvm.polyglot.Value;

final class ShortestDistances<V, W> implements Comparator<V> {
  private final Map<V, W> distances = new HashMap<V, W>();
  private static Value clz =
      ContextInitializer.getPythonClass("/shortestpath/ShortestDistances.py", "ShortestDistances");
  private Value obj;

  public ShortestDistances(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  public void setWeight(V vertex, W distance) {
    //
    // distances.put(vertex, distance);
    //

    obj.invokeMember("setWeight", vertex, distance);
  }

  public W getWeight(V vertex) {
    //
    // return distances.get(vertex);
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("getWeight", vertex).as(W.class);
  }

  public int compare(V left, V right) {
    //
    // if (!alreadyVisited(left) && !alreadyVisited(right)) {
    // return 0;
    // } else if (!alreadyVisited(left)) {
    // return 1;
    // } else if (!alreadyVisited(right)) {
    // return -1;
    // }
    // return weightOperations.compare(getWeight(left), getWeight(right));
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("compare", left, right).as(int.class);
  }

  public boolean alreadyVisited(V vertex) {
    //
    // return distances.containsKey(vertex);
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("alreadyVisited", vertex).as(boolean.class);
  }

  public ShortestDistances(OrderedMonoid<W> weightOperations) {
    //
    // this.weightOperations = weightOperations;
    //

    this.obj = clz.invokeMember("__init__", weightOperations);
  }
}

File: data/skeletons/commons-graph/src/main/org/apache/commons/graph/shortestpath/ShortestDistances.py
# Imports Begin
from src.main.org.apache.commons.graph.weight.OrderedMonoid import *
import typing

# Imports End


class ShortestDistances(Comparator):

    # Class Fields Begin
    __distances: typing.Dict[typing.Any, typing.Any] = None
    __weightOperations: OrderedMonoid[typing.Any] = None
    # Class Fields End

    # Class Methods Begin
    def setWeight(self, vertex: typing.Any, distance: typing.Any) -> None:
        pass

    def getWeight(self, vertex: typing.Any) -> typing.Any:
        pass

    def compare(self, left: typing.Any, right: typing.Any) -> int:
        pass

    def alreadyVisited(self, vertex: typing.Any) -> bool:
        pass

    def __init__(self, weightOperations: OrderedMonoid[typing.Any]) -> None:
        pass

    # Class Methods End
